common/io.cpp

```diff
@@ -234,6 +234,17 @@ ExpectedSize StreamReader::Read(vector<uint8_t>::iterator start, vector<uint8_t>
 	return is_->gcount();
 }
 
+expected::ExpectedSize FileSize(const string &path) {
+	// Probably not as efficient as stat(), but portable.
```

Comment:
Windows? 😄

But seriously, I don't really know. I think Zephyr has optional support for a file system? I just prefer to stick to C++11 if possible, and we can always optimize this later.

Reply:
[Windows is covered](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/stat-functions?view=msvc-170).

I'm just leaning towards using `stat()` where possible and fallback to something less efficient, but potentially more portable where `stat()` is not available. Just like I suggested with `syncfs()`.

Reply:
I think portability is more important than performance currently, at least until a performance problem has been demonstrated. One important reason is that we want to try to code-cover as many lines as possible in Linux, where we can test most effectively. So if it's possible, it's better if they share code.

Syncfs is different, because there is no alternative available in C++11.

What do you think, @oleorhagen and @lluiscampos?

Reply:
I vote for leaning on the `stdlib` everywhere we can :+1:

Reply:
I don't see performance being a problem anywhere in our client atm, except for I/O, and this we haven't spent much time on either.

Reply:
What bugs me more here is that we have some (potentially fake/artificial) suspicion that `stat()` won't be available somewhere and so we don't use it. That's not a good practice in general.

In this particular case, there might be other issues because there is quite a bit of a difference between reading FS metadata and opening a file and seeking to its end. A trivial example is when this code is given a (named) pipe, for example. But there could be many more.

Reply:
Ok, let's the drop the C++11 variant then.

But oh man, once I started pulling at this thread, a lot of things unraveled. I started by looking at C++17, which we have said is ok to use for our POSIX platforms. And they have `filesystem::file_size()`. In fact, they have everything that `boost::filesystem` has! So we don't need `boost::filesystem` at all! We still depend on it indirectly because of Boost Log, but we can remove all our own references and replace with C++17.

Reply:
See https://github.com/mendersoftware/mender/pull/1286.